crypto/rand/rand_meth.c

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2011-2020 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/evp.h>
+#include <openssl/rand.h>
+#include "rand_local.h"
+
+/* Implements the default OpenSSL RAND_add() method */
+static int drbg_add(const void *buf, int num, double randomness)
+{
+    EVP_RAND_CTX *drbg = RAND_get0_primary(NULL);
+
+    if (drbg == NULL || num <= 0)
+        return 0;
+
+    return EVP_RAND_reseed(drbg, 0, NULL, 0, buf, num);
+}
+
+/* Implements the default OpenSSL RAND_seed() method */
+static int drbg_seed(const void *buf, int num)
+{
+    return drbg_add(buf, num, num);
+}
+
+/* Implements the default OpenSSL RAND_status() method */
+static int drbg_status(void)
+{
+    EVP_RAND_CTX *drbg = RAND_get0_primary(NULL);
+
+    if (drbg == NULL)
+        return 0;
+
+    return  EVP_RAND_state(drbg) == EVP_RAND_STATE_READY ? 1 : 0;
+}
+
+/* Implements the default OpenSSL RAND_bytes() method */
+static int drbg_bytes(unsigned char *out, int count)
+{
+    EVP_RAND_CTX *drbg = RAND_get0_public(NULL);
+
+    if (drbg == NULL)
+        return 0;
+
+    return EVP_RAND_generate(drbg, out, count, 0, 0, NULL, 0);
+}
```

Comment:
Yes, good idea.  Additional data isn't intended to increase security, just variety.

Reply:
The additional data collection code lives in the provider now, so it would make sense to do this inside the provider.  Perhaps have the DRBG `reseed` and `generate` calls gather additional input if the passed pointer is NULL?

Reply:
Alternative, an `EVP_RAND_bytes()` call could be introduced that uses the additional data.  I've been tempted to add this API because it would be simpler than using `EVP_RAND_generate()`.  Given that almost everyone won't (& shouldn't) use EVP_RAND directly, I don't see a compelling reason for it.

Reply:
Yes, having `EVP_RAND_bytes()` back (formerly: `RAND_DRBG_bytes()`) would be a good idea.

I would even restore the complete semantics, i.e., move the loop that splits large output buffers back from `EVP_RAND_generate()` to `EVP_RAND_bytes()`. Then `EVP_RAND_generate()` corresponds exactly to to `PROV_generate()` again. (I could restore the original test for the upper limit in drbgtest then).

We can leave this for a separate pr, though.

Reply:
I forgot to say it explicitly: I prefer your second proposal to the first.